File: chundakadan/chundakadan/page/stock_qty/stock_qty.js
frappe.pages['stock-qty'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Stock Quantity Details',
        single_column: true
    });

    let filter_html = `
        <div class="stock-qty-filters row mb-3">
            <div class="col-sm-3 mb-2">
                <select class="form-control" id="so_filter">
                    <option value="">Sales Order ID</option>
                </select>
            </div>
            <div class="col-sm-3 mb-2">
                <input type="date" class="form-control" id="from_date">
            </div>
            <div class="col-sm-3 mb-2">
                <input type="date" class="form-control" id="to_date">
            </div>
            <div class="col-sm-3">
                <button class="btn btn-primary w-100" id="apply_filters"><i class="fa fa-filter"></i> Apply Filters</button>
            </div>
        </div>
    `;
    $(page.body).append(filter_html);

    $(page.body).append(`
        <style>
        .stock-qty-filters select, 
        .stock-qty-filters input, 
        .stock-qty-filters button {
            border-radius: 4px;
        }
        .stock-qty-container table {
            font-size: 13px;
            background: #fff;
            margin-top: 10px;
        }
        .stock-qty-container th {
            background: #f8fafc;
            font-weight: 700;
        }
        .stock-qty-container td, .stock-qty-container th {
            padding: 6px !important;
        }
        </style>
        <div class="stock-qty-container">
            <div id="stock_qty_data"></div>
        </div>
    `);

    // Populate Sales Order dropdown on load
    frappe.call({
        method: 'chundakadan.chundakadan.page.stock_qty.stock_qty.get_sales_order_list',
        callback: function(r) {
            if(r.message && $.isArray(r.message)) {
                const so_filter = $("#so_filter");
                r.message.forEach(so => {
                    so_filter.append(`<option value="${so.name}">${so.name}</option>`);
                });
            }
        }
    });

    // Bind filter button click
    $('#apply_filters').on('click', function() {
        get_and_display_data();
    });

    // Initial load
    get_and_display_data();
};

function get_and_display_data() {
    let filters = {
        sales_order: $('#so_filter').val(),
        from_date: $('#from_date').val(),
        to_date: $('#to_date').val()
    };

    frappe.call({
        method: 'chundakadan.chundakadan.page.stock_qty.stock_qty.get_so_details',
        args: { filters: filters },
        callback: function(r) {
            display_data(r.message || []);
        }
    });
}

function display_data(data) {
    let html = `
        <table class="table table-bordered table-sm">
            <thead>
                <tr>
                    <th>Sales Order</th>
                    <th>Status</th>
                    <th>Customer</th>
                    <th>Date</th>
                    <th>Item Code</th>
                    <th>Ordered Qty</th>
                    <th>Warehouse</th>
                    <th>Stock Qty</th>
                    <th>Remaining Qty</th>
                    <th>Remaining Qty to Invoice</th>
                </tr>
            </thead>
            <tbody>
    `;
    if (!data.length) {
        html += `<tr><td colspan="10" class="text-center text-muted">No Records Found</td></tr>`;
    } else {
        data.forEach(row => {
            html += `
                <tr>
                    <td>${row.sales_order}</td>
                    <td>${row.status || ''}</td>
                    <td>${row.customer}</td>
                    <td>${frappe.datetime.str_to_user(row.date)}</td>
                    <td>${row.item_code}</td>
                    <td>${row.ordered_qty}</td>
                    <td>${row.warehouse}</td>
                    <td>${row.stock_qty}</td>
                    <td>${row.remaining_qty}</td>
                    <td>${row.qty_to_invoice}</td>
                </tr>
            `;
        });
    }
    html += '</tbody></table>';
    $('#stock_qty_data').html(html);
}

File: chundakadan/chundakadan/page/stock_qty/stock_qty.py
import frappe
from frappe import _

@frappe.whitelist()
def get_sales_order_list():
    return frappe.get_all(
        'Sales Order',
        filters={'docstatus': 1},
        fields=['name'],
        order_by='creation desc',
        limit_page_length=100
    )

@frappe.whitelist()
def get_so_details(filters=None):
    import json
    if isinstance(filters, str):
        filters = json.loads(filters)
    filters = filters or {}

    so_filters = {'docstatus': 1}
    if filters.get("sales_order"):
        so_filters["name"] = filters['sales_order']
    if filters.get("from_date") and filters.get("to_date"):
        so_filters["transaction_date"] = ["between", [filters["from_date"], filters["to_date"]]]
    elif filters.get("from_date"):
        so_filters["transaction_date"] = [">=", filters["from_date"]]
    elif filters.get("to_date"):
        so_filters["transaction_date"] = ["<=", filters["to_date"]]

    data = []
    sales_orders = frappe.get_all(
        'Sales Order',
        filters=so_filters,
        fields=['name', 'customer', 'transaction_date', 'status']
    )
    for so in sales_orders:
        so_items = frappe.get_all('Sales Order Item',
            filters={'parent': so.name},
            fields=['name', 'item_code', 'qty', 'warehouse']
        )
        if so_items:
            so_item_names = [item['name'] for item in so_items]
        else:
            so_item_names = []

        invoiced_qty_map = {}
        if so_item_names:
            invoice_rows = frappe.db.sql("""
                SELECT soi.name AS so_item, 
                    IFNULL(SUM(sii.qty), 0) AS billed_qty
                FROM `tabSales Invoice Item` sii
                JOIN `tabSales Order Item` soi ON sii.so_detail = soi.name
                WHERE soi.name IN %(so_item_names)s
                    AND sii.docstatus = 1
                GROUP BY soi.name
            """, {"so_item_names": tuple(so_item_names)}, as_dict=1)
            invoiced_qty_map = {row.so_item: row.billed_qty for row in invoice_rows}

        for item in so_items:
            stock_qty = get_stock_qty(item.item_code, item.warehouse)
            billed_qty = float(invoiced_qty_map.get(item.name, 0))
            qty_to_invoice = float(item.qty) - billed_qty
            remaining_qty = float(stock_qty) - float(item.qty)
            # Only include if qty_to_invoice > 0
            if qty_to_invoice > 0:
                item_data = {
                    'sales_order': so.name,
                    'customer': so.customer,
                    'date': so.transaction_date,
                    'status': so.status,
                    'item_code': item.item_code,
                    'ordered_qty': item.qty,
                    'warehouse': item.warehouse,
                    'stock_qty': stock_qty,
                    'remaining_qty': remaining_qty,
                    'qty_to_invoice': qty_to_invoice,
                }
                data.append(item_data)
    return data

def get_stock_qty(item_code, warehouse):
    bin_data = frappe.db.get_value(
        'Bin',
        {'item_code': item_code, 'warehouse': warehouse},
        ['actual_qty'],
        as_dict=1
    )
    return float(bin_data.actual_qty) if bin_data and bin_data.actual_qty else 0.0
